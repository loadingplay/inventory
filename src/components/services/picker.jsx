class PickerStore {
  constructor() {
    if(!PickerStore.instance) {
      let storage_data;

      try {
        storage_data = JSON.parse(localStorage.getItem('picker-data'));
      } catch(ex) {
        console.log(ex);
      }

      this.data = storage_data ? storage_data:{};
      PickerStore.instance = this;
    }

    return PickerStore.instance;
  }

  set = (sku, value) => {
    this.data[sku] = value;
    localStorage.setItem('picker-data', JSON.stringify(this.data));
  }

  get = (sku) => {
    let number = parseInt(this.data[sku]);
    if (isNaN(number)) return 0;
    return number;
  }

}

const instance = new PickerStore();
Object.freeze(instance);

export default instance;
