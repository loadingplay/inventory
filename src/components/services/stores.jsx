export default class StoreLoader {
  constructor(access_token) {
    this.access_token = access_token;
  }

  loadStoreList = async (page, cellar_id) => {
    let response, json_data;

    response = await fetch(
      `https://replenishments.loadingplay.com/replenishment?items=100&page=${page}&cellar_id=${cellar_id}`,
      {
        "headers": {
          "Authorization": `Bearer ${this.access_token}`
        }
      });

    json_data = await response.json();
    return json_data;
  }
}
