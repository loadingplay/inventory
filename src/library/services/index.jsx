export * from "./picker";
export * from "./stock";
export * from "./stores";
