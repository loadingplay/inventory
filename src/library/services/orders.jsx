
export class Orders {

  static API_URL = "https://apibodegas.loadingplay.com";

  constructor(access_token) {
    this.access_token = access_token;
  }

  create = async (post_data) => {
    let response,
      json_data;

    response = await fetch(
      Orders.API_URL + "/v1/order?",
      {
        method: "POST",
        headers: {
          "Authorization": "Bearer " + this.access_token,
          'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: this._convertToQueryString(post_data)
      }
    );

    json_data = await response.json();
    return json_data
  }

  update = async (order_id, put_data) => {
    let response,
      json_data;

    response = await fetch(
      Orders.API_URL + "/v1/order/" + order_id,
      {
        method: "PUT",
        headers: {
          "Authorization": "Bearer " + this.access_token,
          'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: this._convertToQueryString(put_data)
      }
    );

    json_data = await response.json();
    return json_data;
  }

  get = async (order_id) => {
    let response,
      json_data;

    response = await fetch(
      Orders.API_URL + '/v1/order/' + order_id,
      {
        method: "GET",
        headers: {
          "Authorization": `Bearer ${this.access_token}`
        }
      }
    );

    json_data = await response.json();
    return json_data
  }

  _convertToQueryString = (json_data) => {
    return Object.keys(json_data).map(key => key + '=' + json_data[key]).join('&')
  }
}
