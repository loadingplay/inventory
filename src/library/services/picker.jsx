class _PickerStore {
  constructor() {
    if(!_PickerStore.instance) {
      let storage_data;

      try {
        if (typeof window !== 'undefined' && window.localStorage)
          storage_data = JSON.parse(localStorage.getItem('picker-data'));
      } catch(ex) {
        console.log(ex);
      }

      this.data = storage_data ? storage_data:{};
      _PickerStore.instance = this;
    }

    return _PickerStore.instance;
  }

  set = (cellar_id, sku, value, hq_inventory, current_inventory, suggested) => {
    if (!this.data || !(cellar_id in this.data) || this.data.cellar_id === "") {
      this.data[cellar_id] = {};
    }
    if (value === 0)
      delete this.data[cellar_id][sku];
    else
      this.data[cellar_id][sku] = {
        value,
        hq_inventory,
        current_inventory,
        suggested
      };

    if (typeof window !== 'undefined' && window.localStorage)
      localStorage.setItem('picker-data', JSON.stringify(this.data));
  }

  get = (cellar_id, sku) => {
    if (!this.data || !(cellar_id in this.data)) return 0;

    if (sku === undefined) return this.data[cellar_id];
    if (this.data[cellar_id][sku] === undefined) return 0;

    let number = parseInt(this.data[cellar_id][sku].value);
    if (isNaN(number)) return 0;

    return number;
  }

  clear = (cellar_id) => {
    if (!this.data || !(cellar_id in this.data)) return;
    this.data[cellar_id] = undefined;
    delete this.data[cellar_id];
    if (typeof window !== 'undefined' && window.localStorage)
      localStorage.setItem('picker-data', JSON.stringify(this.data));
  }

  getAll = () => {
    return this.data;
  }

}

export const PickerStore = new _PickerStore();
