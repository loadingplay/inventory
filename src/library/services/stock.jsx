var Loader = function (access_token) {
  this.access_token = access_token;
  this.instances = 0;
  this.max_instances = 15;
  this.chunk_size = 10;
};

Loader.prototype.load = function (cellar_id_list, sku_list) {
  var loader = this;
  cellar_id_list.forEach(function (cellar) {
    loader.loadCellar(cellar, sku_list);
  });
  return this;
};

Loader.prototype.loadCellar = function (cellarid, sku_list) {
  var i, j, temparray;
  for (i = 0, j = sku_list.length; i < j; i += this.chunk_size) {
    temparray = sku_list.slice(i, i + this.chunk_size);
    // do whatever
    this.performRequest(cellarid, temparray);
  }
};

Loader.prototype.performRequest = function (cellarid, skus_chunk) {
  var loader = this;
  if (loader.instances > loader.max_instances) {
    setTimeout(function () {
      loader.performRequest(cellarid, skus_chunk);
    }, 100);
    return
  }
  this.instances += 1;
  fetch("https://apibodegas.loadingplay.com/v1/cellar/" + cellarid + "/product/?sku_list=" + skus_chunk.join(","),
  {
    headers: {
      "Authorization": "Bearer " + this.access_token
    }
  })
  .then((response) => response.json())
  .then((response) => {
    loader.instances -= 1;
    loader.callback(cellarid, response.products);
  });
};


Loader.prototype.done = function (callback) {
  this.callback = callback
};

export function StockLoader(access_token) {
  return new Loader(access_token);
}
