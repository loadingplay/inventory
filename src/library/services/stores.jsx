export class StoreLoader {
  constructor(access_token) {
    this.access_token = access_token;
  }

  loadCellars = async () => {
    let response,
      json_data;

    response = await fetch("https://apibodegas.loadingplay.com/v1/cellar", {
      headers: {
        "Authorization": `Bearer ${this.access_token}`
      }
    });
    json_data = await response.json();
    return json_data;
  }

  loadProducts = async (page, cellar_id, sku_filter) => {
    let response, json_data;

    response = await fetch(
      `https://replenishments.loadingplay.com/replenishment?items=100&page=${page}&cellar_id=${cellar_id}&search_word=${sku_filter}`,
      {
        "headers": {
          "Authorization": `Bearer ${this.access_token}`
        }
      });

    json_data = await response.json();
    return json_data;
  }

  setSelectedCellar = (cellar) => {
    localStorage.setItem("selected-cellar", JSON.stringify(cellar));
  }

  getSelectedCellar = () => {
    return JSON.parse(localStorage.getItem("selected-cellar"));
  }
}
