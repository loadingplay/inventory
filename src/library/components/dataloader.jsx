import React from 'react';
import { StoreLoader, PickerStore, StockLoader } from '../../library/services';

export function withDataLoader(WrappedComponent) {
    return class DataLoader extends React.Component {
      constructor(props) {
        super(props);

        this.access_token = "";
        if (typeof window !== 'undefined' && window.localStorage)
            this.access_token = window.localStorage.getItem("access_token");
        this.picker_data = [];
        this.selected_cellar = {};
        this.hq_cellar = {};
        this.picker_items = [];
        this.store_api = new StoreLoader(this.access_token);
        this.state = { api_items: [], loading_message: "Cargando..." };
        this.getDataFromPicker();
      };

      componentDidMount = async () => {
        await this.getDataFromAPI();
        await this.populateFromKardex();
        this.populateWithPicker();
      }

      getDataFromPicker = () => {
        // get data from picker
        this.picker_data = PickerStore.getAll();
        this.selected_cellar = this.store_api.getSelectedCellar();
        this.hq_cellar = this.store_api.getHQCellar();
        this.picker_items = this.picker_data[this.selected_cellar.id] === undefined ?
          []:this.picker_data[this.selected_cellar.id];
      }

      getDataFromAPI = async () => {
        // get data from API
        this.setState({ loading_message: "Cargando todos los productos sugeridos..." });
        let page = 1,
          api_aux = await this.store_api.searchProducts(page, this.selected_cellar.id, "");

        while (api_aux.status === "success" && api_aux.replenishments.length > 0)
        {
          this.state.api_items = this.state.api_items.concat(api_aux.replenishments.map((item) => {
            return {
              sku: item.sku,
              codigo: item.barcode,
              insumo: item.product_name,
              existencia: {
                casaMatriz: null,
                local: null
              },
              hecho: 0,
              falta: item.suggested
            }
          }));
          api_aux = await this.store_api.searchProducts(++page, this.selected_cellar.id, "");
        }
      }

      populateFromKardex = async () => {
        // populate data from kardex
        let cellars,
          skus,
          stock_loader = StockLoader(this.access_token),
          promises,
          processed_promises = 0,
          done_percentage = 0;

        cellars = this.selected_cellar.id === this.hq_cellar.id ?
          [this.selected_cellar.id] : [this.selected_cellar.id, this.hq_cellar.id]
        skus = this.state.api_items.map((item) => {
            return item.sku;
          });

        stock_loader.max_instances = 1000;
        promises = stock_loader
          .load(cellars, skus)
          .done((loaded_cellar, products) => {
            done_percentage = Math.ceil(++processed_promises * 100 / promises.length);
            this.setState({ loading_message: `Cargando inventario actual ${done_percentage}% ...` });

            products.forEach((y_item) => {
              this.state.api_items.find((x_item) => {
                if (y_item.product_sku !== x_item.sku) return false;

                if (this.hq_cellar.id === loaded_cellar)
                  x_item.existencia.casaMatriz = y_item.balance_units;
                if (this.selected_cellar.id === loaded_cellar)
                  x_item.existencia.local = y_item.balance_units;

                return true;
              });
            });
          });

        await Promise.all(promises);
      }

      populateWithPicker = () => {
        // populate with picker
        this.state.api_items.forEach((item) => {
          if (!this.picker_items[item.sku]) return;

          item.hecho = this.picker_items[item.sku].value;
          item.falta = this.picker_items[item.sku].suggested - this.picker_items[item.sku].value
        });

        // update api items
        this.setState({ api_items: this.state.api_items, loading_message: "" });
      }

      render = () => {
        return <WrappedComponent
          {...this.props}
          items={this.state.api_items}
          loadingMessage={this.state.loading_message}
          selectedCellar={this.selected_cellar}
        />
      }
    }
  }
