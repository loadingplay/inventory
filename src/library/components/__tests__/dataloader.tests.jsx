import React from "react";
import { shallow } from 'enzyme';

import { withDataLoader } from "../dataloader";
import { PickerStore } from '../../services';

describe("Orders", () => {

  beforeEach(jest.clearAllMocks);

  class Wrapped extends React.Component {
    render = () => {
      return <div></div>;
    }
  }

  const DataLoader = withDataLoader(Wrapped);

  test("it should match snapshot", () => {
    const wrapper = shallow(<DataLoader />, { disableLifecycleMethods: true });

    expect(wrapper).toMatchSnapshot('default');
  });

  test("it should load data from picker", () => {
    const wrapper = shallow(<DataLoader />, { disableLifecycleMethods: true });
    const instance = wrapper.instance();

    instance.store_api.getSelectedCellar = jest.fn().mockReturnValue({ id: "3" });
    PickerStore.getAll = jest.fn().mockReturnValue({ "3": "test" });

    instance.getDataFromPicker();

    expect(instance.store_api.getSelectedCellar).toHaveBeenCalledWith();
    expect(instance.picker_items).toEqual("test");
  });

  test("it should get data from API", async () => {
    const wrapper = shallow(<DataLoader />, { disableLifecycleMethods: true });
    const instance = wrapper.instance();

    instance.store_api.searchProducts = jest.fn().mockImplementation((page) => {
      if (page === 1)
        return {status: "success", replenishments: [{
          sku: "test sku",
          barcode: "test barcode",
          product_name: "test product_name",
          suggested: "test suggested"
        }]};

      return {};
    })
    instance.selected_cellar = { id: "test" }

    await instance.getDataFromAPI();

    expect(instance.state.loading_message).toEqual("Cargando todos los productos sugeridos...");
    expect(instance.store_api.searchProducts).toHaveBeenCalledWith(1, "test", "");
    expect(instance.state.api_items).toEqual([
      {
        "codigo": "test barcode",
        "existencia": {
          "casaMatriz": null,
          "local": null,
        },
        "falta": "test suggested",
        "hecho": 0,
        "insumo": "test product_name",
        "sku": "test sku"
      }
    ]);
  });
});
