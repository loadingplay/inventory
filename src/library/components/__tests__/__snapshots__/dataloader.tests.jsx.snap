// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders it should match snapshot: default 1`] = `
<Wrapped
  items={Array []}
  loadingMessage="Cargando..."
  selectedCellar={Object {}}
/>
`;
